helm-v3/006-repositories.md

```diff
@@ -1,40 +1,153 @@
+- A repository v2 specification will be defined, and the index file will be
+JSON (`index.json` vs `index.yaml`)
+
+## Capabilities API
+
+The Capabilties API is designed to provide Helm with information such as:
+
+1. Does this domain support actions such as login or push?
+2. What supported method should the client use to perform a specific action?
+3. What metadata does the client need to initiate the action?
+
+This will allow Helm to support any number of things in the future related
+to repositories without making breaking changes (e.g. remote `helm search`).
+
+### `/.well-known/helm`
+
+The Capabilities API will be enabled for a specific domain via
+[Well-Known URI](https://tools.ietf.org/html/rfc5785) located at
+`/.well-known/helm`.
+
+*Note: This file may also serve to provide configuration for other APIs
+in the future by supplying an additional top-level section.*
+
+The response body should be valid JSON and should supply a valid,
+top-level `capabilities` section. For example:
+
+```
+{
+    "capabilities": {
+        "login": {
+            ...
+        },
+        "push": {
+            ...
+        },
+        "search": {
+            ...
+        }
+    }
+}
+```
+
+Each subsection (i.e. `login`, `push`, `search`) will provide configuration
+specific to the capabiltity.
+
+For example, the following might inform Helm that this domain only
+accepts token-based login:
+
+```
+...
+        "login": {
+            "methods": {
+                "token": {
+                    "loginUrl": "https://auth.site.com/oauth2/authorize",
+                    "tokenUrl": "https://auth.site.com/oauth2/token"
+                }
+            }
+        }
+...
+```
+
+## The `helm login` command
+
+A new `helm login` command will be added to Helm, responsible for
+managing authentication across repository providers.
+
+The basic usage will be:
+
+```
+helm login <url>
+```
+
+This command is intended to allow users to authenticate just one time in
+order to access multiple repositories provided by the same source. 
+
+Afterwards, running a `helm repo add` pointing to any repository URL
+tied to a provider that has been successfully authenticated against will
+not require additional credentials.
+
+The exact semantics are TBD, but several auth methods will be supported,
+such as:
+
+* Basic Auth
+* Cert based Auth
+* Token/Bearer Auth (with token refresh)
+* Two-Factor Auth
+
+## The `helm push` command
+
+A new `helm push` command will be added to Helm, responsible for uploading
+chart versions to a repository.
+
+The basic usage will be:
+
+```
+helm push <chart> <repo>
+```
+
+The exact semantics of this command are still TBD.
+
+
+The most basic push method supported is referred to as `http`, and should be
+present in `capabilities` in order for push to be enabled:
+
+```
+...
+        "push": {
+            "methods": {
+                "http": {
+                    "method": "POST",
+                    "path": "/api/charts"
```

Comment:
if the repository's upload API path is `POST https://hub.charts.io/charts/{namespace}`, and I push chart like this:
``` bash
helm push mychart-1.0.1.tgz https://hub.charts.io/mynamespace
```
what the capabilities API's push path would like, how can I represent the namespace path variable in capabilities API 's response? I assume that `{path}` represent the path of push url, then the capabilities API's push config will look like this:
```
...
        "push": {
            "methods": {
                "http": {
                    "method": "POST",
                    "path": "/charts/{path}"
                }
            }
        }
...
```
Is that right? But what if the upload API path is `POST https://hub.charts.io/charts/{namespace}/{repo}` ?